src/cli/commands/run.js

```diff
@@ -118,6 +119,13 @@ export async function run(config: Config, reporter: Reporter, flags: Object, arg
     }
 
     if (cmds.length) {
+      const ignoreEngines = !!(flags.ignoreEngines || config.getOption('ignore-engines'));
+      try {
+        await checkCompatibility({_reference: {}, ...pkg}, config, ignoreEngines);
+      } catch (err) {
+        throw err instanceof MessageError ? new MessageError(reporter.lang('cannotRunWithIncompatibleEnv')) : err;
```

Comment:
This refines a, in this context, vague [error](https://github.com/yarnpkg/yarn/blob/4eb6391efbbe5f3dbce3c5c4f31c246c7130c7da/src/package-compatibility.js#L161).